Sources/ApolloCodegenLib/IR/IR.swift

```diff
@@ -62,30 +62,87 @@ class IR {
   /// Multiple `SelectionSet`s may select fields on the same `Entity`. All `SelectionSet`s that will
   /// be selected on the same object share the same `Entity`.
   class Entity {
-    struct FieldPathComponent: Hashable {
-      let name: String
-      let type: GraphQLType
+
+    /// Represents the location within a GraphQL definition (operation or fragment) of an `Entity`.
+    struct Location: Hashable {
```

Comment:
In order to generate selection set names properly, an entity needs to be able to indicate where it is located in its definition. This includes needing to know _what_ definition (operation or fragment) it is defined in.

Previously, we were using a hacky solution of adding an initial `FieldPath` component that included the name of the fragment, or the operation type. This was being used to help properly generate the selection set names, but it was hacky for 2 reasons: 

1. It was making the "field path" not really accurate. Field path included a first component for the source definition that any code that was consuming the field path just needed to understand was there. Causing tight coupling of implementation code.
2. This partially put the responsibility of how we generate names in the templates into the IR. The IR should just represent the shape of the definitions, and the templates consume them to understand how to generate the correct code.

Representing the `Location` with a `SourceDefinition` and a `FieldPath` allows us to better represent the shape of the operations in the IR and allows the templates to do all the logic for formatting names of the Selection Sets.

Reply:
Like this change, especially helping separate responsibilities between the IR and Templates